mixnet/node/src/lib.rs

```diff
@@ -77,205 +75,244 @@ impl MixnetNode {
+    ) -> Self {
+        Self {
+            message_tx,
+            message_rx,
+            connections: HashMap::with_capacity(config.connection_pool_size),
+            config,
+        }
+    }
+
+    pub async fn run(mut self) {
+        loop {
+            tokio::select! {
+                msg = self.message_rx.recv() => {
+                    if let Some(msg) = msg {
+                        self.handle_msg(msg).await;
+                    } else {
+                        // Channel closed, we should shutdown the packet forwarder thread
+                        return;
+                    }
+                },
+                _ = tokio::signal::ctrl_c() => {
+                    tracing::info!("Shutting down packet forwarder thread...");
+                    return;
+                }
+            }
+        }
+    }
+
+    async fn try_send(&mut self, target: SocketAddr, body: &Body) -> Result<()> {
+        if let std::collections::hash_map::Entry::Vacant(e) = self.connections.entry(target) {
+            match TcpStream::connect(target).await {
+                Ok(tcp) => {
+                    e.insert(tcp);
+                }
+                Err(e) => {
+                    tracing::error!("failed to connect to {}: {e}", target);
+                    return Err(MixnetNodeError::Protocol(e.into()));
+                }
+            }
+        }
+        body.write(self.connections.get_mut(&target).unwrap())
             .await
-            .map_err(Into::into);
+            .map_err(From::from)
+    }
+
+    async fn handle_msg(&mut self, msg: Message) {
+        self.send(msg).await
+    }
+
+    async fn send(&mut self, msg: Message) {
+        if let Err(err) = self.try_send(msg.target, &msg.body).await {
+            match err {
+                MixnetNodeError::Protocol(ProtocolError::IO(e))
+                    if e.kind() == std::io::ErrorKind::Unsupported =>
+                {
+                    tracing::error!("fail to send message to {}: {e}", msg.target);
+                }
+                _ => self.handle_retry(msg),
+            }
+        }
+    }
+
+    fn handle_retry(&self, mut msg: Message) {
+        if msg.retry_count < self.config.max_retries {
+            let delay = Duration::from_millis(
+                (self.config.retry_delay.as_millis() as u64).pow(msg.retry_count as u32),
+            );
+            let tx = self.message_tx.clone();
+            tokio::spawn(async move {
+                tokio::time::sleep(delay).await;
+                msg.retry_count += 1;
+                if let Err(e) = tx.send(msg) {
+                    tracing::error!("fail to enqueue retry message: {e}");
+                }
+            });
+        } else {
+            tracing::error!(
+                "fail to send message to {}: reach maximum retries",
+                msg.target
+            );
+        }
+    }
+}
+
+pub struct Message {
+    target: SocketAddr,
+    body: Body,
+    retry_count: usize,
+}
```

Comment:
In the mixnet-related crates, I would like to suggest to have clearly separated terminology for 'message' and 'packet'. As we've seen so far, we call the object, that the mixnet end-user wants to send, "message". The message is splitted into "packets" by the mixclient, and those packets are sent to mixnodes. The mixnode forwards packets to other mixnodes, not messages (Only one additional case is that the mixnode also forwards final payloads to the destination mixnode).

That is, we don't need to use the terminology "message" in the mixnode layer. In this PR, these two or three terminologies are mixed. This will cause confusions in the future.

Reply:
Yeah, the thread is PacketHandler but the struct is "message", which is misleading. I change all "message" to "packet".